scenarios/two_party_countersigning/src/main.rs

```diff
@@ -117,132 +119,109 @@ fn agent_behaviour_initiate(
-                                            continue;
-                                        }
-                                        Err(_) => {
-                                            // Must be resilient to unexpected signals, somebody else might try to initiate with us while we're already
-                                            // working with another peer.
-                                            log::debug!("Got an unexpected signal, will try again. {:?}", signal);
-                                            continue;
-                                        }
-                                    };
-
-                                    log::debug!("The other party [{:?}] has accepted the countersigning session.", agent_pub_key);
-
-                                    let retry_count = complete_session(client.clone(), app_port, cell_id.clone(), my_preflight_response, other_response, session_timeout).await.context("Initiator failed to complete session")?;
-
-                                    let elapsed = start.elapsed();
-
-                                    log::debug!("Completed the countersigning session with agent {:?}", agent_pub_key);
-
-                                    let initiated_success = initiated_success.fetch_add(1, std::sync::atomic::Ordering::Acquire);
-                                    reporter.add_custom(
-                                        ReportMetric::new("countersigning_session_initiated_success")
-                                            .with_tag("agent", agent_name.clone())
-                                            .with_field("retries", retry_count as u64)
-                                            .with_field("value", initiated_success as u64),
-                                    );
-                                    reporter.add_custom(
-                                        ReportMetric::new("countersigning_session_initiated_duration")
-                                            .with_tag("agent", agent_name)
-                                            .with_field("value", elapsed.as_secs_f64()),
-                                    );
-
-                                    break;
-                                }
-                                None => {
-                                    log::warn!("No signal received, problem with the remote? Will try again.");
-                                }
-                            }
-                        }
+                    let session_times = my_preflight_response.request.session_times.clone();
+                    let session_timeout = Instant::now().add(Duration::from_millis(
+                        (session_times.end.as_millis() - session_times.start.as_millis()) as u64,
+                    ));
 
-                        // Add no new agents, that should only happen when we exhaust the list.
-                        Ok(Vec::with_capacity(0))
+                    match run_initiated_session(
```

Comment:
Hideous diff, the code to the left and above has been moved to a function. The reason is to make sure that any errors that happen after the session has started won't cause the agent behavior to continue to the next iteration. We have to wait for the session to time out if something goes wrong!